### PR TITLE
Add unsafe-deserialization support for Jabsorb

### PR DESCRIPTION
**Pull request by smowton**
This is partly extracted from https://github.com/github/codeql/pull/5954

### CODE REVIEW

File: java/ql/src/semmle/code/java/security/UnsafeDeserializationQuery.qll [hunk 1]
```diff
@@ -243,9 +251,36 @@ class UnsafeDeserializationConfig extends TaintTracking::Configuration {
...
+ * and its name contains "resolve", "load", etc.
+ *
+ * Any method call that satisfies the rule above is assumed to propagate taint from its string arguments,
+ * so methods that accept user-controlled data but sanitize it or use it for some
+ * completely different purpose before returning a type descriptor could result in false positives.
+ */
+predicate looksLikeResolveClassStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+  exists(MethodAccess ma, Method m, Expr arg | m = ma.getMethod() and arg = ma.getAnArgument() |
+    m.getReturnType() instanceof JacksonTypeDescriptorType and
+    m.getName().toLowerCase().regexpMatch("(?i).*(resolve|load|class|type).*") and
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeDeserializationQuery.qll:274]:
```suggestion
    m.getName().regexpMatch("(?i).*(resolve|load|class|type).*") and
```

File: java/ql/src/semmle/code/java/frameworks/Jackson.qll [hunk 2]
```diff
@@ -77,14 +77,6 @@ class SetPolymorphicTypeValidatorSource extends DataFlow::ExprNode {
   }
 }
 
-/** Holds if `fromNode` to `toNode` is a dataflow step that resolves a class. */
-predicate resolveClassStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
-  exists(ReflectiveClassIdentifierMethodAccess ma |
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/Jackson.qll:82]: Did this removal make any imports superfluous in this file?

**Review by aschackmull — COMMENTED** [review]
A few comments, otherwise LGTM.

File: java/ql/src/semmle/code/java/frameworks/Jackson.qll (hunk 2 above)
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/Jackson.qll:82]: Yes, removed one

**Comment by smowton** [comment]
@aschackmull done

**Review by aschackmull — APPROVED** [review]

File: java/ql/src/semmle/code/java/security/UnsafeDeserializationQuery.qll [hunk 3]
```diff
@@ -256,7 +291,12 @@ class UnsafeTypeConfig extends TaintTracking2::Configuration {
...
   override predicate isSink(DataFlow::Node sink) {
     exists(MethodAccess ma, int i, Expr arg | i > 0 and ma.getArgument(i) = arg |
-      ma.getMethod() instanceof ObjectMapperReadMethod and
+      (
+        ma.getMethod() instanceof ObjectMapperReadMethod
+        or
+        ma.getMethod() instanceof JabsorbUnmarshallMethod
+      ) and
+      // Note `JacksonTypeDescriptorType` includes plain old `java.lang.Class`
       arg.getType() instanceof JacksonTypeDescriptorType and
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/security/UnsafeDeserializationQuery.qll:300]: Should `JacksonTypeDescriptorType` now be refactored as well given that the predicates using it are not exclusively for Jackson anymore?

**Pull request merged**